content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md

```diff
@@ -0,0 +1,380 @@
+  image     = "nvidia" # or "ubuntu"
+
+  # cloud-specific config
+  cloud     = "aws"    # see `git checkout generic` branch for: gcp, az, k8s
+
+  # blank means extract from local env vars
+  environment = { NGROK_TOKEN = "", TF_CPP_MIN_LOG_LEVEL = "1", QUIET = "1", GITHUB_USER = "username" }
+  storage {
+    workdir = "shared"
+    output  = "."
+  }
+```
+
+The second part of the configuration is perhaps the most interesting. Here we
+provide the specifications for the instance Terraform should provision for us.
+In this case, we would like a medium CPU and an NVIDIA T4 GPU, along with a
+125GB disk. Because machine types vary between cloud vendors,
+[TPI does some translation](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type)
+from generic types (e.g., `s`/`m`/`l`/`xl`) to specific cloud machine types.
+This allows us to generalize these configurations and quickly switch from AWS to
+Azure, for example.
+
+Of particular interest here is the `spot = 0`, which tells TPI to provision spot
+instances instead of on-demand ones. At the time of writing, the on-demand
+hourly rate for this type of instance is $0.526. On the other hand, spot
+instances are only $0.15/hour. With little to no effort, TPI allows us to reduce
+our cloud costs by 70%!
+
+<admon type="tip">
+
+We could also set an upper limit to the $/hr price we are willing to pay (e.g.,
+`spot = 0.12`), and TPI would hold off on provisioning until instances are
+available at that price. Disabling spot pricing (`spot = -1`) would make TPI
+provision on-demand instances.
+
+</admon>
+
+After specifying the details for the hardware, we configure a few environment
+variables for the instance. We also set the local working directory for the
+script and the directory to download files from (relative to the working
+directory) upon terminating the instance. In this case, both are the `shared`
+directory.
+
+```bash
+#!/bin/bash
+set -euo pipefail
+if test "$GITHUB_USER" != username; then
+  # SSH debugging
+  trap 'echo script error: waiting for debugging over SSH. Run \"terraform destroy\" to stop waiting; sleep inf' ERR
+  mkdir -p "$HOME/.ssh"
+  curl -fsSL "https://github.com/$GITHUB_USER.keys" >> "$HOME/.ssh/authorized_keys"
+fi
+export CUDACXX=/usr/local/cuda/bin/nvcc
+export DEBIAN_FRONTEND=noninteractive
+sed -ri 's#^(APT::Periodic::Unattended-Upgrade).*#\1 "0";#' /etc/apt/apt.conf.d/20auto-upgrades
+dpkg-reconfigure unattended-upgrades
+# install dependencies
+pip3 install $${QUIET:+-q} jupyterlab notebook matplotlib ipywidgets tensorflow==2.8.0 tensorboard tensorflow_datasets
+(curl -fsSL https://deb.nodesource.com/setup_16.x | bash -) >/dev/null
+apt-get install -y $${QUIET:+-qq} nodejs
+
+# start tunnel
+export JUPYTER_TOKEN="$(uuidgen)"
+pushd "$(mktemp -d --suffix dependencies)"
+npm i ngrok
+npx ngrok authtoken "$NGROK_TOKEN"
+(node <<TUNNEL
+const fs = require('fs');
+const ngrok = require('ngrok');
```

Comment:
This whole code section is where I get pretty much completely lost. While there are some comments put in which is helpful, there are other parts that are not explained.  What is CUDACXX? DEBIAN_FRONTEND?

Will this all look happy and understandable to a data scientist used to their Jupyter Notebook?  Or do we need to do some more work here?
Is this script interchangeable for all use cases?  What parts will need to be changed depending on their needs?

Reply:
I feel we don't need to dump the whole script here. Could condense it a lot. For a start:

```suggestion
#!/bin/bash
# install dependencies
pip install jupyterlab notebook matplotlib ipywidgets tensorflow tensorboard tensorflow_datasets
apt-get install -yq nodejs
npm i ngrok

# start tunnel
export JUPYTER_TOKEN="$(uuidgen)"
npx ngrok authtoken "$NGROK_TOKEN"
(node <<TUNNEL
const fs = require('fs');
const ngrok = require('ngrok');
```

Reply:
So I guess this needs to be corrected too. ☝️